File: Serial OrderWise/LeetCode 992 - Subarrays with K Different Integers/Solution.java
class Solution {
    private int solve(int[] arr, int k) {
        int n = arr.length;
        int l = 0, r = 0;
        Map<Integer, Integer> map = new HashMap<>();
        map.put(arr[0], 1);
        int ans = 0;
        while(l <= r) {
            if(map.size() <= k) {
                ans += r-l+1;
                if(r < n-1) {
                    r++;
                    map.put(arr[r], map.getOrDefault(arr[r], 0) + 1);
                } else {
                    break;
                }
            } else {
                map.put(arr[l], map.getOrDefault(arr[l],0)-1);
                if(map.get(arr[l]) <= 0) {
                    map.remove(arr[l]);
                }
                l++;
            }
        }
        return ans;
    }
    public int subarraysWithKDistinct(int[] nums, int k) {
        return solve(nums, k) - solve(nums, k-1);
    }
}

File: Serial OrderWise/LeetCode 992 - Subarrays with K Different Integers/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    int solve(vector<int>& nums, int k) {
        unordered_map<int, int> map;
        int l = 0, r = 0, n = nums.size();
        map[nums[0]] = 1;
        long ans = 0;
        while(l<=r) {
            if(map.size()<=k) {
                ans += r - l + 1;
                if(r < n-1) {
                    r++;
                    map[nums[r]]++;
                } else {
                    break;
                }
            } else {
                map[nums[l]]--;
                if(map[nums[l]] <= 0) {
                    map.erase(nums[l]);
                }
                l++;
            }
        }
        return ans;
    }
public:
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        return solve(nums, k) - solve(nums, k-1);
    }
};